services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+    q = "SELECT secret FROM oauth_client WHERE id = ?"
+
+insertOAuthAuthorizationCode :: (MonadClient m) => Word64 -> OAuthAuthorizationCode -> OAuthClientId -> UserId -> OAuthScopes -> RedirectUrl -> m ()
+insertOAuthAuthorizationCode ttl code cid uid scope uri = do
+  let cqlScope = C.Set (Set.toList (unOAuthScopes scope))
+  retry x5 . write q $ params LocalQuorum (code, cid, uid, cqlScope, uri)
+  where
+    q :: PrepQuery W (OAuthAuthorizationCode, OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl) ()
+    q = fromString $ "INSERT INTO oauth_auth_code (code, client, user, scope, redirect_uri) VALUES (?, ?, ?, ?, ?) USING TTL " <> show ttl
+
+lookupOAuthAuthorizationCode :: (MonadClient m) => OAuthAuthorizationCode -> m (Maybe (OAuthClientId, UserId, OAuthScopes, RedirectUrl))
+lookupOAuthAuthorizationCode code = do
+  mTuple <- retry x5 . query1 q $ params LocalQuorum (Identity code)
+  pure $ mTuple <&> \(cid, uid, C.Set scope, uri) -> (cid, uid, OAuthScopes (Set.fromList scope), uri)
+  where
+    q :: PrepQuery R (Identity OAuthAuthorizationCode) (OAuthClientId, UserId, C.Set OAuthScope, RedirectUrl)
+    q = "SELECT client, user, scope, redirect_uri FROM oauth_auth_code WHERE code = ?"
+
+deleteOAuthAuthorizationCode :: (MonadClient m) => OAuthAuthorizationCode -> m ()
+deleteOAuthAuthorizationCode code = retry x5 . write q $ params LocalQuorum (Identity code)
+  where
+    q :: PrepQuery W (Identity OAuthAuthorizationCode) ()
+    q = "DELETE FROM oauth_auth_code WHERE code = ?"
+
+lookupAndDeleteOAuthAuthorizationCode :: (MonadClient m) => OAuthAuthorizationCode -> m (Maybe (OAuthClientId, UserId, OAuthScopes, RedirectUrl))
+lookupAndDeleteOAuthAuthorizationCode code = lookupOAuthAuthorizationCode code <* deleteOAuthAuthorizationCode code
+
+insertOAuthRefreshToken :: (MonadClient m) => Word32 -> Word64 -> OAuthRefreshTokenInfo -> m ()
+insertOAuthRefreshToken maxActiveTokens ttl info = do
+  let rid = oriId info
+  oldTokes <- determineOldestTokensToBeDeleted <$> lookupOAuthRefreshTokens (oriUserId info)
+  for_ oldTokes deleteOAuthRefreshToken
+  retry x5 . write qInsertId $ params LocalQuorum (oriUserId info, rid)
+  retry x5 . write qInsertInfo $ params LocalQuorum (rid, oriClientId info, oriUserId info, C.Set (Set.toList (unOAuthScopes (oriScopes info))), oriCreatedAt info)
+  where
+    qInsertInfo :: PrepQuery W (OAuthRefreshTokenId, OAuthClientId, UserId, C.Set OAuthScope, UTCTime) ()
+    qInsertInfo = fromString $ "INSERT INTO oauth_refresh_token (id, client, user, scope, created_at) VALUES (?, ?, ?, ?, ?) USING TTL " <> show ttl
+
+    qInsertId :: PrepQuery W (UserId, OAuthRefreshTokenId) ()
+    qInsertId = fromString $ "INSERT INTO oauth_user_refresh_token (user, token_id) VALUES (?, ?) USING TTL " <> show ttl
+
+    determineOldestTokensToBeDeleted :: [OAuthRefreshTokenInfo] -> [OAuthRefreshTokenInfo]
+    determineOldestTokensToBeDeleted tokens =
+      take (length sorted - fromIntegral maxActiveTokens + 1) sorted
```

Comment:
```
length sorted === 13
maxActiveTokens === 10
```

then this would yield the oldes `13 - 10 + 1 === 4` tokens, but it should yield 3, no?

Reply:
alternatively, we could reject creating the new tokens.  i would be less surprised by an error here if i were an oauth client.

Reply:
Let's say:
- max active tokens is 10
- there are currently 13 tokens (however that happened)
- user wants to create a new token
- then we need to delete the 4 oldest tokens so that after adding the new token we have a max of 10